### PR TITLE
feat: support no POI

### LINKED ISSUE #514: background-only workspaces cannot be built as models
# Description

If we specify a background-only workspace, this passes the `workspace.json` schema, but we cannot build a Model from it, but we can build a Workspace from it.

# Expected Behavior

We expect to be able to build a (background-only?) Model.

# Actual Behavior

An error message like

```
pyhf.exceptions.InvalidModel: The parameter of interest 'mu_SIG' cannot be fit as it is not declared in the model specification.
```

# Steps to Reproduce

Using the background-only JSON from the sbottom analysis causes this. Then running

```python
import pyhf, json
w = pyhf.Workspace(json.load(file('BkgOnly.json')))
w.model()
```

# Checklist

- [x] Run `git fetch` to get the most up to date version of `master`
- [x] Searched through existing Issues to confirm this is not a duplicate issue
- [x] Filled out the Description, Expected Behavior, Actual Behavior, and Steps to Reproduce sections above or have edited/removed them in a way that fully describes the issue

### LINKED ISSUE #946: allow poi-less Models
# Description

the `.set_poi()` call in the `Model` ctor should be optional. It's only necessary for hypotests, but e.g. you can do MLE fits without problems w/o a POI

It's also useful when teaching modifiers etc.

### PR DESCRIPTION
**Pull request by lukasheinrich**
# Description

closes #514, closes #946 

```
* Allow POI-less models (poi_name = None)
* Raise exception when used in contexts that require POI
```

### CODE REVIEW

File: src/pyhf/infer/mle.py [hunk 1]
```diff
@@ -76,6 +76,10 @@ def fixed_poi_fit(poi_val, data, pdf, init_pars=None, par_bounds=None, **kwargs)
         See optimizer API
 
     """
+    if not pdf.config.poi_index:
+        raise RuntimeError(
```
Comment (lukasheinrich) [inline src/pyhf/infer/mle.py:80]: should this a pyhf-specific exception?

File: src/pyhf/infer/test_statistics.py [hunk 2]
```diff
@@ -42,6 +42,11 @@ def qmu(mu, data, pdf, init_pars, par_bounds):
     Returns:
         Float: The calculated test statistic, :math:`q_{\mu}`
     """
+    if not pdf.config.poi_index:
+        raise RuntimeError(
```
Comment (lukasheinrich) [inline src/pyhf/infer/test_statistics.py:46]: dito

File: src/pyhf/infer/mle.py (hunk 1 above)
Comment (kratsg) [reply src/pyhf/infer/mle.py:80]: yes

File: src/pyhf/infer/test_statistics.py (hunk 2 above)
Comment (kratsg) [reply src/pyhf/infer/test_statistics.py:46]: yes

File: tests/test_pdf.py [hunk 3]
```diff
@@ -5,6 +5,26 @@
 import json
 
 
+def test_hello_world_ctor():
```
Comment (matthewfeickert) [inline tests/test_pdf.py:8]: I think I'm tired, but what is `ctor`?

File: tests/test_pdf.py [hunk 4]
```diff
@@ -182,6 +202,39 @@ def test_pdf_integration_staterror(backend):
...
+    m = pyhf.Model(spec, poi_name=None)
+
+    data = [12] + m.config.auxdata
+    pyhf.infer.mle.fit(data, m)
+
+    with pytest.raises(RuntimeError):
+        pyhf.infer.mle.fixed_poi_fit(1.0, data, m)
+
+    with pytest.raises(RuntimeError):
+        pyhf.infer.hypotest(1.0, data, m)
```
Comment (matthewfeickert) [inline tests/test_pdf.py:235]:
```suggestion
    model = pyhf.Model(spec, poi_name=None)

    data = [12] + model.config.auxdata
    pyhf.infer.mle.fit(data, model)

    with pytest.raises(RuntimeError):
        pyhf.infer.mle.fixed_poi_fit(1.0, data, model)

    with pytest.raises(RuntimeError):
        pyhf.infer.hypotest(1.0, data, model)
```

Nitpicky nitpick of preferring variable names being longer than 1 letter.

File: tests/test_pdf.py (hunk 3 above)
Comment (lukasheinrich) [reply tests/test_pdf.py:8]: constructor

Comment (matthewfeickert) [inline tests/test_pdf.py:8]: Maybe more explicit naming?
```suggestion
def test_minimum_model_spec():
```

File: tests/test_pdf.py [hunk 5]
```diff
@@ -182,6 +202,39 @@ def test_pdf_integration_staterror(backend):
...
+    m = pyhf.Model(spec, poi_name=None)
+
+    data = [12] + m.config.auxdata
+    pyhf.infer.mle.fit(data, m)
+
+    with pytest.raises(pyhf.exceptions.UnspecifiedPOI):
+        pyhf.infer.mle.fixed_poi_fit(1.0, data, m)
+
+    with pytest.raises(pyhf.exceptions.UnspecifiedPOI):
+        pyhf.infer.hypotest(1.0, data, m)
```
Comment (matthewfeickert) [inline tests/test_pdf.py:235]:
```suggestion
    model = pyhf.Model(spec, poi_name=None)

    data = [12] + model.config.auxdata
    pyhf.infer.mle.fit(data, model)

    with pytest.raises(pyhf.exceptions.UnspecifiedPOI):
        pyhf.infer.mle.fixed_poi_fit(1.0, data, model)

    with pytest.raises(pyhf.exceptions.UnspecifiedPOI):
        pyhf.infer.hypotest(1.0, data, model)
```

For consistency with the rests of the test I'd stil suggest no single letters vars.

File: tests/test_pdf.py (hunk 3 above)
Comment (matthewfeickert) [reply tests/test_pdf.py:8]: Thanks. :+1: Much more obvious now after sleep, my bad.

**Pull request merged**